snapcraft/plugins/go.py

```diff
@@ -93,15 +103,22 @@ def get_pull_properties(cls):
 
     def __init__(self, name, options, project):
         super().__init__(name, options, project)
-        if "go" not in self.build_snaps and not any(
-            (s.startswith("go/") for s in self.build_snaps)
-        ):
-            self.build_packages.append("golang-go")
+
+        self._setup_base_tools(options.go_channel, project.info.base)
+
         self._gopath = os.path.join(self.partdir, "go")
         self._gopath_src = os.path.join(self._gopath, "src")
         self._gopath_bin = os.path.join(self._gopath, "bin")
         self._gopath_pkg = os.path.join(self._gopath, "pkg")
 
+    def _setup_base_tools(self, go_channel, base):
+        if go_channel != "":
+            self.build_snaps.append("go/{}".format(go_channel))
+        elif base in ("core16", "core18"):
+            self.build_packages.append("golang-go")
+        else:
+            raise errors.PluginBaseError(plugin_name=self.name, base=base)
```

Comment:
To be clear: you anticipate this plugin will run fine on any base as long as the Go snap is being used?

Reply:
I anticipate it is a good starter for anyone trying to start a port to a new device.